language/move-binary-format/src/deserializer.rs

```diff
@@ -43,9 +43,21 @@ impl CompiledModule {
         binary: &[u8],
         max_binary_format_version: u32,
     ) -> BinaryLoaderResult<Self> {
-        let module = deserialize_compiled_module(binary, max_binary_format_version)?;
-        BoundsChecker::verify_module(&module)?;
-        Ok(module)
+        let prev_state = move_core_types::state::set_state(VMState::DESERIALIZER);
+        let result = std::panic::catch_unwind(|| {
+            let module = deserialize_compiled_module(binary, max_binary_format_version)?;
+            BoundsChecker::verify_module(&module)?;
+
+            Ok(module)
+        })
+        .unwrap_or_else(|_| {
+            Err(PartialVMError::new(
+                StatusCode::VERIFIER_INVARIANT_VIOLATION,
```

Comment:
can we make a specific error instead of a `VERIFIER_INVARIANT_VIOLATION`?
`VERIFIER_INVARIANT_VIOLATION` is raised in other context as well and its handling may vary depending on the app embedding the VM.
I think this deserves its own error that would be easy to filter out from a client which could do the proper thing.
For instance I would imagine a client would want to raise an alert in this cases.
In that respect I even wonder if this code should live on the client and we should have clear entry points (which I think we do). I guess filtering a specific error may be fine and a client may have all the info in place to log something helpful.
Thoughts on this?

Reply:
This error code looks good to me as long as we attach a message. Its used in many places for similar reason, and we should have monitoring and alerting on for it. Besides we need to keep those PRs minimal please for cherry picking.